src/keystores/base_keystore.rs

```diff
@@ -0,0 +1,333 @@
+use crate::{
+    consts::UNICODE_CONTROL_CHARS,
+    errors::KeystoreError,
+    sk_to_pk,
+    utils::{pbkdf2, scrypt_key},
+};
+use aes::{
+    cipher::{generic_array::GenericArray, KeyIvInit, StreamCipher},
+    Aes128,
+};
+use ctr::Ctr128BE;
+use num_traits::ToPrimitive;
+use rand::Rng;
+use serde::{Deserialize, Serialize};
+use serde_json::{Map, Value};
+use sha2::{Digest, Sha256};
+use std::io::Write;
+use std::{collections::HashMap, fs, os::unix::fs::PermissionsExt};
+use unicode_normalization::UnicodeNormalization;
+use uuid::Uuid;
+
+#[derive(Serialize, Deserialize, Debug, Default, PartialEq)]
+pub struct KeystoreModule {
+    pub function: String,
+    #[serde(default)]
+    pub params: HashMap<String, serde_json::Value>,
+    pub(crate) message: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Default, PartialEq)]
+pub struct KeystoreCrypto {
+    pub kdf: KeystoreModule,
+    pub(crate) checksum: KeystoreModule,
+    pub cipher: KeystoreModule,
+}
+
+impl KeystoreCrypto {
+    fn from_json(json_dict: &Map<String, Value>) -> Result<Self, KeystoreError> {
+        let kdf: KeystoreModule =
+            serde_json::from_value(json_dict["kdf"].clone()).map_err(|e| KeystoreError::from(e))?;
+        let checksum: KeystoreModule = serde_json::from_value(json_dict["checksum"].clone())?;
+        let cipher: KeystoreModule = serde_json::from_value(json_dict["cipher"].clone())?;
+        Ok(Self {
+            kdf,
+            checksum,
+            cipher,
+        })
+    }
+}
+
+/// Keystore is an EIP 2335-compliant keystore. A keystore is a JSON file that
+/// stores an encrypted version of a private key under a user-supplied password.
+///
+/// Ref: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2335.md
+#[derive(Serialize, Deserialize, Debug, Default, PartialEq)]
+pub struct Keystore {
+    pub crypto: KeystoreCrypto,
+    pub(crate) description: String,
+    pub(crate) pubkey: String,
+    pub path: String,
+    pub(crate) uuid: String,
+    pub(crate) version: u32,
+}
+
+impl Keystore {
+    fn get_u32(param: Option<Value>) -> Result<u32, KeystoreError> {
```

Comment:
can you add some docs to these methods?

Reply:
Was thinking actually to create a different PR with the docs.